### ISSUE TITLE
Missing [assembly: AssemblyMetadata("Serviceable", "True")]

### ISSUE THREAD

**Issue by bricelam** [dotnet/arcade#1526]
If a project is serviceable an can potentially run on .NET Framework (e.g. targeting net472, netstandard2.0, etc.) it should include the following assembly-level attribute.

``` csharp
[assembly: AssemblyMetadata("Serviceable", "True")]
```

cc @Eilon

**Comment by Eilon** [comment]
Here's an old thread where this is also discussed: https://github.com/dotnet/cli/issues/3345

**Comment by markwilkie** [comment]
Thanks @bricelam - good timing as we'll be tackling servicing of Arcade soon.

**Comment by tmat** [comment]
@bricelam 
How is this related to Arcade? Are you suggesting that we add a target that adds this attribute to all assemblies built with Arcade?

**Comment by bricelam** [comment]
Yes, add a target to Arcade that adds this attribute to the assembly if the MSBuild `Serviceable` property is `ture`. KoreBuild did it [here](https://github.com/aspnet/BuildTools/blob/release/2.2/src/Internal.AspNetCore.Sdk/build/GenerateAssemblyInfo.targets#L30-L51).

**Comment by Eilon** [comment]
Yeah the attribute is supposed to be present in all DLLs that might run on .NET FX. It causes a breadcrumb trail to be left (in the registry?) so that WU/MU can scan and see if a DLL was ever used on the machine, in case it needs to be serviced.

**Comment by vatsan-madhavan** [comment]
I just noticed this as well. `System.Windows.Forms.dll`, `System.Xaml.dll` etc. are produced from an Arcade Sdk based repo, and do not contain this attribute. Whereas assemblies like `WindowsBase.dll` that are produced in a buildtools based repo contain the following attributes: 

```
[assembly: AssemblyMetadata(".NETFrameworkAssembly", "")]
[assembly: AssemblyMetadata("PreferInbox", "True")]
[assembly: AssemblyMetadata("Serviceable", "True")]
```

These attributes are populated in BuildTools [here](https://github.com/dotnet/buildtools/blob/28606b8791ccf507651febbf41c480dd87d4a507/src/Microsoft.DotNet.Build.Tasks/PackageFiles/Build.Common.targets#L38)

Can this be added in Arcade Sdk as well ? 

cc @ericstj, @rladuca, @stevenbrix

**Comment by tmat** [comment]
Should these be actually used for WPF buily for .NET Core?

```C#
[assembly: AssemblyMetadata(".NETFrameworkAssembly", "")]
[assembly: AssemblyMetadata("PreferInbox", "True")]
```

These won't be shipping in desktop .NET Framework, right?

**Comment by vatsan-madhavan** [comment]
I agree that the metadata values generated by buildtools by default, and inherited by WPF today, don't make good sense. 

`.NETFrameworkAssembly` is plain wrong, (maybe it should be something else for .NET Core assemblies?). How is `[AssemblyIdentity.IsFrameworkAssembly`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.tasks.deployment.manifestutilities.assemblyidentity.isframeworkassembly?view=netframework-4.7.2&viewFallbackFrom=netcore-2.2) supposed to function on .NET Core CoreFx assemblies? Is it supposed to return `False` because `.NET Core` <> `.NET Framework`? If so, we should probably not even add the `.NETFrameworkAssembly` metadata value - and perhaps no replacement is necessary. 

I'm not even sure what semantics `PreferInbox` is supposed to convey. 

Perhaps `Serviceable` is the only one we will end up with eventually? If so, who/what is the intended reader of this metadata - do we know ?

**Comment by tmat** [comment]
@Eilon Explained the meaning of Serviceable here: https://github.com/dotnet/arcade/issues/1526#issuecomment-447079742

**Comment by Eilon** [comment]
@tmat said:

> Another question about just the serviceable metadata. Should `[assembly: AssemblyMetadata("Serviceable", "True")]` be by default added to all assemblies that are shipping (`IsShipping=true`) and target framework identifier is `.NET Framework` or `.NET Standard` (with a switch to override this default)?

This seems reasonable to me.

**Comment by markwilkie** [comment]
@Pilchie - given the recent ASP servicing changes, do you know if this is still "a thing"?

**Comment by Pilchie** [comment]
I don't know of any other work that would address it.  We still do have several packages that are .NET Standard, and our other servicing work is only related to reference assemblies, where I assume this needs to be in reference asssemblies.

**Comment by blowdart** [comment]
If assemblies aren't marked, and are targeted at Core, are we still dropping the breadcrumbs needed for hammer patching?

**Comment by ericstj** [comment]
On core the host writes breadcrumbs based on what's in the deps file, not what's in the assembly.  @vitek-karas

**Comment by blowdart** [comment]
Just checking :)

**Comment by dougbu** [comment]
We think the ASP.NET assemblies are fine but an automated confirmation would be worth adding to the Arcade infrastructure. @jcagme is this already done in the validation ring? If not, could it be❔

In particular, could this be combined w/ signing validation (where the assemblies are already available) @joeloff❔

**Comment by joeloff** [comment]
It makes sense to add a validation step in the build. I'd image it could be an inline task, unless you want something like an assembly validation tool that can extract assemblies from other containers, e.g. zip, nupkg, installer package, etc. Sign validation mostly runs on larger containers (as those need to be signed too). I'm unsure whether it's being executed on the raw assemblies when they are signed.
